mm/pmm.c

```diff
@@ -108,7 +128,19 @@ static frames_array_t *new_frames_array(void) {
     if (!boot_flags.virt)
         array = (frames_array_t *) mfn_to_virt_kern(frame->mfn);
     else {
-        array = vmap_kern_4k(mfn_to_virt_map(frame->mfn), frame->mfn, L1_PROT);
+        /* switch to special refilling mode to avoid deadlock with paging */
+        ASSERT(!refilling);
+        refilling = true;
+
+        /* only paging will be allowed to take the lock for writing while refilling */
+        spin_unlock(&priority_lock);
+        array = vmap_frame_refill(mfn_to_virt_map(frame->mfn), frame->mfn, from_paging);
+        spin_lock(&priority_lock);
+
+        /* switch back to normal mode */
+        ASSERT(refilling);
+        refilling = false;
```

Comment:
it might make sense to handle all this code as independent functions (see my comment for the `vmap_frame_refill`).